### ISSUE TITLE
Improve Model training

### ISSUE THREAD

**Issue by radeusgd** [Luna-Tensorflow/Luna-Tensorflow#94]
Currently model `train` does just one pass over the training set, that's very rudimentary and far far from what is done in practice.

To make the function at least slightly more usable following features should be added:
- **epochs** - repeat the training on the whole set N times
- **loss history** - return an array of average losses in each epoch
- **validation** - make it possible to sacrifice part of the set for validation, there are two solutions possible (and it would be best if both were supported for example in two versions of the function of with a helper function to do the sampling):
    - provide a validation_split argument that splits the training set into training and validation according to some ratio
    - make it possible to explicitly provide a validation set that is used
- (nice to have) **early stopping epochs** - if loss on validation doesn't decrease for K consequentive epochs, stop training early

**Comment by radeusgd** [comment]
Result of this issue should be changing Model::train function into
```
train :: Int -> List (Tensor, Tensor) -> Validation -> EarlyStoppingConf -> (Model, List Float)
def train epochs data validation earlystop:
    ...
    (newmodel, lossHistory)
```
where
```
class Validation:
    NoValidation
    ValidationPercentage Float
    ValidationSet List (Tensor, Tensor)

class EarlyStoppingConf:
    NoEarlyStopping
    StopAfter Int # stops if for n epochs loss on validation set did not decrease
```

loss_history should include history of average validation loss in each epoch, or train loss if no validation set is provided

**Comment by wojtrz13** [comment]
Related: f407aa0ccb714eb36ce13c0f69957c4c1c0dd47b
